broadinstitute/single_cell_portal_core
# Sort cell filters by count, enable sorting by label (SCP-5373)

This shows more relevant filters at a glance, through useful ordering.  It also lets researchers better group related filters.

### Overview
Previously (#1920), filters were ordered unintuitively, as determined by their incidental order in original data and unrelated SCP processing.  That confused a beta tester.

Now, filters are ordered more usefully.  Filters are sorted by cell count, by default.  This makes it effortless to discern the _distribution_ of the various categories that filters represent.  The physical size difference of more digits helps visualize spreads.  Having the most numerous category at top also makes it easiest to create a big visual difference in the plot, because the group with the most cells -- sometimes _a lot_ more cells than the others -- becomes most easily hidden.

Users can also easily change the default sort-by-count to sort-by-label.  This can display related filters together, e.g. in the biosample ID facet, where A) cell counts are often fairly evenly distributed across filters and B) there is often _internal structure_ in the filter labels ("BM03_13wk_r1", "BM03_13wkpp_r2192", etc.).  Once you see these related filters side-by-side, you see how their counts change as you update filters in other facets.  That uncovers meaningful patterns.

### Video

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/b3e9507c-692f-49f4-9915-b80d1a38e30e

### Possible work future
In the future, it might be worth refining a few things:
* The "Reset filters" button could reset not just filter selections, but also filter sorting.  I briefly tried this, and it ended up being more complex than I anticipated, so this seems sensible to defer.

* The sort-by-label implementation is sometimes awkward.  The rudimentary [natural sort](https://en.wikipedia.org/wiki/Natural_sort_order) is often only natural in a technical sense; in the more relevant and common sense, it sometimes orders things unnaturally.  This is particularly apparent in biosample IDs for SCP1671.  Likewise, annotations are sometimes Boolean-esque with values like "yes", "no", "NA".  These cases could probably be easily inferred and cast with more useful orderings as seen in publications, to avoid less conventional orders like "no", "NA", "yes".

### Test
A new automated test verifies this new functionality.  To manually test:

1.  Go to "Human milk - differential expression" study
2. Click "Filter plotted cells"
3. Click "More..." in the "Cell type" facet
4. Confirm filters are ordered by count
5. Click the sort icon at top right in facet
6. Confirm filters are ordered by label

This satisfies SCP-5373.
